src/main/java/org/wildfly/security/auth/spi/Attributes.java

```diff
@@ -576,13 +598,69 @@ default boolean isEmpty() {
     }
 
     /**
-     * Get the collection of values for the given key.  The result may implement {@link SetEntry} if the values
-     * are distinct (for example, a role or group set).
+     * Returns a read-only instance of this instance.
      *
-     * @param key the attribute name
-     * @return the (possibly empty) attribute value collection
+     * @return a read-only instance of this instance.
      */
-    Entry get(String key);
+    default Attributes asReadOnly() {
+        return new Attributes() {
+            private Map<String, Entry> entryCache = new HashMap<>();
+            private Collection<Entry> cachedEntries;
+
+            @Override
+            public Collection<Entry> entries() {
+                if (this.cachedEntries != null) {
+                    return this.cachedEntries;
+                }
+
+                return this.cachedEntries = new AbstractCollection<Entry>() {
+                    Set<String> keySet = Attributes.this.keySet();
```

Comment:
You don't have to cache the keySet here; you can fetch it anew each time (the outer instance should cache it).